platforms/software/dependency-management/src/test/groovy/org/gradle/internal/component/local/model/DefaultProjectComponentSelectorTest.groovy

```diff
@@ -76,10 +80,26 @@ class DefaultProjectComponentSelectorTest extends Specification {
 
     def "matches id (#buildName #projectPath)"() {
         expect:
-        def selector = new DefaultProjectComponentSelector(new ProjectIdentity(Stub(BuildIdentifier), Path.path(":id:path"), Path.path(":project:path"), "projectName"), ImmutableAttributes.EMPTY, [])
+        def selector = new DefaultProjectComponentSelector(new ProjectIdentity(Stub(BuildIdentifier), Path.path(":id:path"), Path.path(":project:path"), "projectName"), ImmutableAttributes.EMPTY, ImmutableSet.of())
         def sameIdPath = new DefaultProjectComponentIdentifier(Stub(BuildIdentifier), Path.path(":id:path"), Path.path(":project:path"), "projectName")
         def differentIdPath = new DefaultProjectComponentIdentifier(Stub(BuildIdentifier), Path.path(":id:path2"), Path.path(":project:path"), "projectName")
         selector.matchesStrictly(sameIdPath)
         !selector.matchesStrictly(differentIdPath)
     }
+
+    def "capabilities are exposed"() {
+        def capabilities = ImmutableSet.of(
+            new DefaultSpecificCapabilitySelector(new DefaultImmutableCapability("org", "blah", "1")),
+            new DefaultFeatureCapabilitySelector("foo")
+        )
+        def identity = new ProjectIdentity(Stub(BuildIdentifier), Path.path(":id:path"), Path.path(":project:path"), "projectName")
+        ProjectComponentSelector selector = new DefaultProjectComponentSelector(identity, ImmutableAttributes.EMPTY, capabilities)
+
+        expect:
+        selector.capabilitySelectors == capabilities
+
+        selector.requestedCapabilities.size() == 1 // We do not have access to the project instance and cannot derived a capability for the feature selector.
```

Comment:
❓ Can you explain this more?  

Looking at this test makes me think something has gone wrong.  It would be perhaps making this test work with 2 `DefaultSpecificCapabilitySelector`s, and making a copy with a more specific name explaining that it is verifying the silent ignoring of the `DefaultFeatureCapabilitySelector` (and why).

Reply:
Yeah this is probably the one downfall of this change. 

A `ProjectComponentSelector` does not have a reference to the backing project. It only has a `ProjectIdentity` (and should not have a `Project` reference). The identity contains:
- The project path relative to the build
- The project path relative to the build tree
- The project name
- The build path

The project name, group, version are all mutable and are not known at the time of creating the selector. So, to ask the for "resolved" capability from the selector doesn't really make sense at this point. All we know is the requested feature, and we cannot know the "resolved" capability until this component selector is resolved to a project and we know that projects name, group, and version. 

Can you explain your second paragraph a bit more? My intention of this test _was_ to verify that we are silently ignoring the `DefaultFeatureCapabilitySelector`. I updated the test to explain the _why_ a bit more (as I did in this comment)

Reply:
What I meant is: this test adds 2 capability selectors to `capabilities`, which is then used to create `selector`.  Then it checks that `selected.requestedCapabilities.size == 1`.  Where did the `DefaultFeatureCapabilitySelector` go?  Why is it not returned?

So I'd take this current test, and replace `DefaultFeatureCapabilitySelector` with another `DefaultSpecificCapabilitySelector`.  And I'd keep an (unchanged) copy that explains why the feature selector disappears - which is what your current comment does.  That way there would be one (simple) test that verifies that `capabilities are exposed` and one `only specific capabilities are available` (or something) containing the commented behavior.

Reply:
I can split this into two tests